### PR TITLE
v0.9.1: auth, metrics, and server extensions

### PR DESCRIPTION
**Pull request by daniel-nichter**
### New
* Add `metrics/` (GET /metrics): group-based metrics for every route with an entity type (i.e. all the routes minus metadata routes like GET /metrics)
  * Add `etre.EntityClient.WithTrace(string)` to provide trace metrics
  * 86% - 93% test coverage
* Add `auth/`: conventional RBAC implemented via plugin interface (default plugin allows all)
  * 100% test coverage
* Add `server/` and `app/`: boot loader pattern: move startup code/logic out of main.go and into these so 3rd party wrappers can provide `app.Context.Plugins.Auth`, for example

### API changes
* Refactor pre- and post-route middleware in `api.API` to accommodate metrics, auth, etc.
* Consolidate entity validation and write op (`wo`) in pre-route rather than copy-pasting same check in every controller

### Config changes
Makes it easy for new users/devs to get started with Etre by making it work with zero config, straight out of the box
* Add `acl` section to config
* Consolidate `CDCConfig`:
  * Remove top-level `Delay DelayConfig`
  * Move `DelayConfig.Collection` to `CDCConfig.DelayCollection`  (yaml: `delay_collection`)
* Change `tls-*` yaml tags to `tls_*` (`_` is the norm)
* Remove `ServerConfig.UsernameHeader` (automatically uses `auth.Caller.Name`)
* Make `-config` optional: built-in defaults are sufficient to run (default entity type: `host`)
* Change default config file from `"/etc/etre/etre.yaml` to `config/<env>.yaml`
* Config file only required to exist and load if given explicitly (`-config`). New default ^ config file loads if possible; else, if not, Etre uses built-in defaults.

### CDC changes
* Add `cdc.NoopStore` and `cdc.NoopDelayer` when cdc is disabled (instead of using test/mock objects)

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+	// 3rd-party extensions, all optional
+	Hooks   Hooks
+	Plugins Plugins
+}
+
+type Hooks struct {
+	LoadConfig func(Context) (config.Config, error)
+	RunAPI     func() error
+	StopAPI    func() error
```
Comment (felixplajer) [inline app/app.go]: A comment explaining RunAPI and StopAPI would be helpful.

File: auth/auth.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"net/http"
+)
+
+type ACL struct {
+	// User-defined role. This must exactly match a Caller role for the ACL
+	// to match.
+	Role string
+
+	// Role grants admin access (all ops) to request. Mutually exclusive with Ops.
```
Comment (felixplajer) [inline auth/auth.go]: "Ops" - is that a field that doesn't exist anymore?

File: auth/auth.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	// Return a new caller each time because the auth manager might set
+	// Trace if the client passed trace values via X-Etre-Trace header
+	caller := Caller{
+		Name:         "etre",
+		MetricGroups: []string{"etre"},
+	}
+	return caller, nil
+}
+
+func (a AllowAll) Authorize(Caller, Action) error {
```
Comment (felixplajer) [inline auth/auth.go]: Would be good to note that returning nil means they're authorized and returning an error means they're not.

Comment (felixplajer) [reply auth/auth.go]: Actually, put that comment up in the Plugin interface, not the AllowAll implementation.

File: auth/manager.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			allowed = true
+			break
+		}
+	}
+	if !allowed {
+		return fmt.Errorf("caller %s has no role that allows %s %s entities; caller roles: %v", caller.Name, opName, a.EntityType, caller.Roles)
+	}
+
+	// Let plugin do final authorization
+	return m.plugin.Authorize(caller, a)
```
Comment (felixplajer) [inline auth/manager.go]: What kind of auth is the plugin going to be doing? It seems like the manager already does everything possible with just the Caller and Action.

File: entity_client.go [hunk 5]
```diff
@@ -53,6 +53,11 @@ type EntityClient interface {
 	// caller must ensure that Set.Size is greater than zero and Set.Op and Set.Id are nonempty
 	// strings.
 	WithSet(Set) EntityClient
+
+	// WithTrace returns a new EntityClient that sends the trace string every request
```
Comment (felixplajer) [inline entity_client.go]: "sends the trace string _with/for_ every request"

File: metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+	EntityType(string)
+
+	// Inc increments the metric name (mn) by n. The metric must be a counter.
+	Inc(mn byte, n int64)
+
+	// IncLabel increments the metric name (mn) counter for the label by 1.
+	// The metric must be LabelRead, LabelUpdate, or LabelDelete.
+	IncLabel(mn byte, label string)
+
+	// IncError increments the metric name (mn) by 1. The metirc must be one
```
Comment (felixplajer) [inline metrics/metrics.go]: `metirc` -> `metric`

File: metrics/metrics.go [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+	// Trace increments trace value counters by 1.
+	Trace(map[string]string)
+
+	// Report returns a snapshot of all metrics, calculating stats like average
+	// and percentiles.
+	Report() etre.MetricsReport
+}
+
+var _ = &metrics{} // ensure metrics implements Metrics
```
Comment (felixplajer) [inline metrics/metrics.go]: I've seen this several times now, but I don't understand - how does this ensure metrics implements Metrics? Wouldn't it need to be something like `var _ (Metrics) = &metrics{}`?

Comment (daniel-nichter) [reply metrics/metrics.go]: Yes, you're correct. Fixed.

File: metrics/metrics.go (hunk 6 above)
Comment (daniel-nichter) [reply metrics/metrics.go]: Fixed.

File: entity_client.go (hunk 5 above)
Comment (daniel-nichter) [reply entity_client.go]: Fixed.

File: auth/auth.go (hunk 3 above)
Comment (daniel-nichter) [reply auth/auth.go]: Fixed.

File: auth/auth.go (hunk 2 above)
Comment (daniel-nichter) [reply auth/auth.go]: No, was copy-paste error. Fixed.

File: app/app.go (hunk 1 above)
Comment (daniel-nichter) [reply app/app.go]: Done.

**Pull request merged**